File: scala/50.scala
object Solution {
    def myPow(x: Double, n: Int): Double = {
        def tailPow(x: Double, n: Int, res: Double): Double = {
            n match {
                case 0 => res
                case 1 => x * res
                case _ => tailPow(x, n/2, res*res)
            }
        }
        if (n > 0) tailPow(x, n, 1)
        else tailPow(1/x, -(n+1), 1/x)/x
    }
}

File: leetcodeJava/50.java
package leetcodeJava;

class Solution {
    public double myPow(double x, int nn) {
        // n to bits
        long n = nn;
        if (n < 0) {
            x = 1/x;
            n = -n;
        }
        if (n == 0) {
            return 1;
        }
        double res = 1, tmp = x;
        while (n != 0) {
            if (n%2 == 1) {
                res = res * tmp;
            } 
            n = n/2;
            tmp = tmp * tmp;
        }
        return res;
    }
}